MapValue clean up

pontusmelke opened the pull request:
We are in many places leaking the internal implementation of MapValue which makes it hard to try out different implementations. This PR removes all leakages of the internal `java.util.Map` of `MapValue`

pontusmelke committed:
Allow for different MapValue implementations

pontusmelke committed:
Add filter function on MapValue

pontusmelke committed:
Add updatedWith to MapValue

pontusmelke committed:
Remove usage of getMapCopy

pontusmelke committed:
Code style and fixes

pontusmelke committed:
Move compine into updateWith

pontusmelke committed:
Remove copy method for MapValues

pontusmelke committed:
Remove map( Map<String, AnyValue) replace with builder

pontusmelke committed:
Tentatively use MutableMap from eclipse collections

pontusmelke committed:
Don't use eclipse collection as map

All benchmarks show that using a normal hash map is 2-3 times faster.

pontusmelke committed:
Fix bug in LazyMap::size

pontusmelke committed:
Fix bug in MapValues::equals

pontusmelke committed:
Update test not to enforce order of `keys`

pontusmelke committed:
typo: columne -> column

pontusmelke committed:
update test mocks

pontusmelke committed:
Update compiled runtime to use the correct map method

SaschaPeukert commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageReader.java:
```diff
@@ -101,8 +100,10 @@ public void read( BoltRequestMessageHandler handler ) throws IOException
+
+        authTokenValue.foreach(
+                ( key, value ) -> tokenMap.put( key, writer.valueAsObject( value ) ) );
+        return tokenMap;
```
This is way more readable! Thanks!

pontusmelke committed:
Forcefully load lazy map for map projection

This used to work accidentally since the `VirtualValues::combine` exhausted
and copied over all maps.

pontusmelke committed:
Fixed bug in PointFunction

PointFunction was in some cases returning a MapValue instead of
the PointValue created from that MapValue

pontusmelke committed:
Fixed bug in gathering notification

pontusmelke committed:
PropertiesFunction must load lazy properties

It used to work because we always made a copy in PropertiesFunction
however this is not necessary in the general case, just for LazyMap

pontusmelke committed:
Disable tck tests

The tests mandates a specific order of `keys` and must be fixed on the
TCK side

SaschaPeukert commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/commands/expressions/DesugaredMapProjection.scala:
```diff
@@ -40,12 +37,18 @@ case class DesugaredMapProjection(id: String, includeAllProps: Boolean, literalE
-    val mapOfLiteralValues = literalExpressions.map {
-      case (k, e) => (k, e(ctx, state))
-    }.toMap.asJava
+    val builder = new MapValueBuilder()
```
```val builder = new MapValueBuilder(literalExpressions.size)```  ?

SaschaPeukert commented on community/cypher/cypher/src/main/java/org/neo4j/cypher/internal/codegen/CompiledMaterializeValueMapper.java:
```diff
@@ -106,9 +105,9 @@ public AnyValue mapRelationship( VirtualRelationshipValue value )
-            Map<String,AnyValue> map = new HashMap<>();
-            value.foreach( ( k, v ) -> map.put( k, v.map( this ) ) );
-            return VirtualValues.map( map );
+            MapValueBuilder builder = new MapValueBuilder();
```
we could initialize this new ```MapValueBuilder``` with ```value.size```, right?

pontusmelke replied:
`value.size()` may be a O(n) operation so I think I prefer not to do it in this case.

SaschaPeukert commented on community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/CypherAdapterStream.java:
```diff
@@ -152,24 +150,26 @@ public static AnyValue convert( Iterable<Notification> notifications )
-                notificationMap.put( "title", stringValue( notification.getTitle() ) );
-                notificationMap.put( "description", stringValue( notification.getDescription() ) );
-                notificationMap.put( "severity", stringValue( notification.getSeverity().toString() ) );
+                MapValueBuilder builder = new MapValueBuilder();
```
```new MapValueBuilder(4);```

pontusmelke replied:
I'm torn in this case it is mostly four but in some cases if `includePosition` is true we include a fifth element, so maybe go with size 5 instead.

SaschaPeukert commented on community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/ExecutionPlanConverter.java:
```diff
@@ -44,21 +43,21 @@ private ExecutionPlanConverter()
-        out.put( "args", ValueUtils.asMapValue( plan.getArguments() ) );
-        out.put( "identifiers", ValueUtils.asListValue( plan.getIdentifiers() ) );
-        out.put( "children", children( plan ) );
+        MapValueBuilder out = new MapValueBuilder(  );
```
Could it make sense to initialize this one with a size?

SaschaPeukert reviewed (commented):
Awesome work, just a few minor comments and you might need to rebase it

fickludd commented on community/values/src/main/java/org/neo4j/values/virtual/MapValueBuilder.java:
```diff
@@ -0,0 +1,52 @@
+
+import org.neo4j.values.AnyValue;
+
+public class MapValueBuilder
```
Wow, using a builder pattern here is great. Awesome.

fickludd commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/MapSupport.scala:
```diff
@@ -84,7 +85,10 @@ class LazyMap[T](ctx: QueryContext, ops: Operations[T], id: Long)
 
   //we need a way forcefully load lazy values
-  def load(): Boolean = allProps != null
+  def load(): MapValue =
```
If we really need lazy values, then maybe we should upgrade `load()` to something like `MapValue.materialize()` so all maps had this functionality. In general I'm afraid of any lazy values and the bugs they might cause. This also applies to lazily filtered and mapped values, where the filtering/mapping function might not give the same result at the time of execution when it's lazy.

fickludd reviewed (commented):
I have only skimmed this, but there are some nice improvements here. In particular like the `MapValueBuilder`. 

I think when the `ListValue` implementations were introduced: There should be some test to assert that all the different `MapValue` implementations are equal (and share hashcode) if they have the same contents, and that they are not equal when they have different contents.

pontusmelke committed:
Fixes from code review

pontusmelke committed:
Better testing of different MapValue implementations

pontusmelke committed:
checkstyle

pontusmelke committed:
More license headers
